app/rtc_compare.py

```diff
@@ -0,0 +1,484 @@
+'''
+Compare two RTC products (in HDF5) if they are equivalent.
+Part of the codes are copied from PROTEUS SAS
+'''
+
+import argparse
+import itertools
+
+import h5py
+import numpy as np
+
+RTC_S1_PRODUCTS_ERROR_TOLERANCE = 1e-6
+
+def _get_parser():
+    parser = argparse.ArgumentParser(
+        description='Compare two RTC products',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+    # Inputs
+    parser.add_argument('input_file',
+                        type=str,
+                        nargs=2,
+                        help='Input RTC products in NETCDF/HDF5 format')
+
+    return parser
+
+
+def _unpack_array(val_in, hdf5_obj_in):
+    '''
+    Unpack the array of array into ordinary numpy array.
+    Convert an HDF5 object reference into the path it is pointing to.
+
+    For internal use in this script.
+
+    Parameter:
+    -----------
+    val_in: np.ndarray
+        numpy array to unpack
+    hdf5_obj_in:
+        Source HDF5 object of `val_in`
+
+    Return:
+    val_out: np.ndarray
+        unpacked array
+
+    '''
+    list_val_in = list(itertools.chain.from_iterable(val_in))
+
+    list_val_out = [None] * len(list_val_in)
+    for i_val, element_in in enumerate(list_val_in):
+        if isinstance(element_in, h5py.h5r.Reference):
+            list_val_out[i_val] = np.str_(hdf5_obj_in[element_in].name)
+        else:
+            list_val_out[i_val] = element_in
+    val_out = np.array(list_val_out)
+
+    return val_out
+
+
+def print_data_difference(val_1, val_2, indent=4):
+    '''
+    Print out the difference of the data whose dimension is >= 1
+
+    Parameters:
+    -----------
+    val_1, val_2: np.array
+        Data that has difference to each other
+    indent: int
+        Number of spaces for indentation
+    '''
+
+    str_indent = ' ' * indent + '-'
+
+    # printout the difference when the array is numeric
+    if issubclass(val_1.dtype.type, np.number):
+        difference_val = val_1 - val_2
+        index_max_diff = np.nanargmax(np.abs(difference_val))
+
+        print(f'{str_indent} Maximum difference detected from index '
+              f'[{index_max_diff}]: '
+              f'1st: ({val_1[index_max_diff]}), 2nd: ({val_2[index_max_diff]}) = '
+              f'diff: ({difference_val[index_max_diff]})')
+    else:
+        # Print out the first discrepancy in case of non-numerical array
+        flag_discrepancy = (val_1 != val_2)
+        index_first_discrepancy = np.where(flag_discrepancy)[0][0]
+        print(f'{str_indent} The first discrepancy has detected from index '
+              f'[{index_first_discrepancy}] : '
+              f'1st=({val_1[index_first_discrepancy]}), '
+              f'2nd=({val_2[index_first_discrepancy]})')
+
+    # Check pixel-by-pixel nan / non-nan difference
+    if (issubclass(val_1.dtype.type, np.floating) or
+        issubclass(val_1.dtype.type, np.complexfloating)):
```

Comment:
It looks like you can "early return" here as well.

Reply:
Maybe it would make more sense to let the "NaN check" to follow the difference printout routine because we take care of all tasks for numerical arrays in one place. See the snipped below:

```python
# printout the difference
    if issubclass(val_1.dtype.type, np.number):
        # Routine for numeric array
        difference_val = val_1 - val_2
        index_max_diff = np.nanargmax(np.abs(difference_val))

        print(f'{str_indent} Maximum difference detected from index '
              f'[{index_max_diff}]: '
              f'1st: ({val_1[index_max_diff]}), 2nd: ({val_2[index_max_diff]}) = '
              f'diff: ({difference_val[index_max_diff]})')

        # Check pixel-by-pixel nan / non-nan difference
        if (issubclass(val_1.dtype.type, np.floating) or
            issubclass(val_1.dtype.type, np.complexfloating)):

            mask_nan_val_1 = np.isnan(val_1)
            mask_nan_val_2 = np.isnan(val_2)

            mask_nan_discrepancy = np.logical_xor(mask_nan_val_1, mask_nan_val_2)

            if not np.any(mask_nan_discrepancy):
                print(f'{str_indent} NaN discrepancy was not detected.')
                return None

            num_pixel_nan_discrepancy = mask_nan_discrepancy.sum()
            index_pixel_nan_discrepancy = np.where(mask_nan_discrepancy)
            print(f'{str_indent} Found {num_pixel_nan_discrepancy} '
                'NaN inconsistencies between input arrays. '
                'First index of the discrepancy: '
                f'[{index_pixel_nan_discrepancy[0][0]}]')
            print(f'{str_indent} val_1[{index_pixel_nan_discrepancy[0][0]}] = '
                f'{val_1[index_pixel_nan_discrepancy[0][0]]}')
            print(f'{str_indent} val_2[{index_pixel_nan_discrepancy[0][0]}] = '
                f'{val_2[index_pixel_nan_discrepancy[0][0]]}')

            # Operations to print out further info regarding the discrapancy
            num_nan_both = np.logical_and(mask_nan_val_1, mask_nan_val_2).sum()
            num_nan_val_1 = np.sum(mask_nan_val_1)
            num_nan_val_2 = np.sum(mask_nan_val_2)
            print(f'{str_indent} # NaNs on val_1 only: {num_nan_val_1 - num_nan_both}')
            print(f'{str_indent} # NaNs on val_2 only: {num_nan_val_2 - num_nan_both}')
```

Removing the if statement `if (issubclass(val_1.dtype.type, np.floating) or ...`  in the snippet above would make the code work anyhow because `np.isnan()` is defined for integers. We can simplify the code but the "NaN analysis" is unnecessary for integers? Let me know what you think.

Reply:
That works, but notice that it's much easier to check if the arrays are numeric. If they are not, you handle it and return right away (everything that comes after doesn't need to be indented). After that, you can check if they are floating, if they are not, you return right away. That will simplify the code.

Reply:
Modified the code accordingly. Thanks for the suggestion!